cache/remote.go

```diff
@@ -104,15 +113,98 @@ func (sr *immutableRef) GetRemote(ctx context.Context, createIfNeeded bool, comp
 			}
 		}
 
-		remote.Descriptors = append(remote.Descriptors, desc)
-		mprovider.Add(lazyRefProvider{
+		layerP := lazyRefProvider{
 			ref:     ref,
 			desc:    desc,
 			dh:      sr.descHandlers[desc.Digest],
 			session: s,
-		})
+		}
+		mu.Lock()
+		mprovider.Add(layerP)
+		mu.Unlock()
+		if forceCompression {
```

Comment:
I feel like this code that creates the blobs with the correct compression type conceptually makes more sense to be in the existing `computeBlobChain` code, somewhere around here maybe: https://github.com/moby/buildkit/blob/edc28d1b028bcf240039134dae3176f174e71130/cache/blobs.go#L57-L58

Not a strong opinion though, so if that's easier said than done I don't think it's a big deal to leave it here.